service-level-sidecar/bucketd.js

```diff
@@ -0,0 +1,65 @@
+const arsenal = require('arsenal');
+const bucketclient = require('bucketclient');
+
+const { BucketClientInterface } = arsenal.storage.metadata.bucketclient;
+const { usersBucket, splitter: mdKeySplitter } = arsenal.constants;
+
+const rootLogger = require('./log');
+const env = require('./env');
+const utils = require('./utils');
+
+const PAGE_SIZE = 1000;
+
+const params = {
+    bucketdBootstrap: [env.bucketd],
+    https: env.bucketdTls ? env.tls.certs : undefined,
+};
+
+const metadata = new BucketClientInterface(params, bucketclient, rootLogger);
+
+const listObjects = utils.retryable(metadata.listObject.bind(metadata));
+
+/**
+ * List all s3 buckets implemented as a async generator
+ *
+ * @param {object} log - werelogs logger instance
+ * @returns {object} - async generator
+ */
+async function* listBuckets(log) {
+    const listingParams = { maxKeys: PAGE_SIZE, listingType: 'Basic' };
+    let gt;
+
+    while (true) {
+        let res;
+
+        try {
+            // eslint-disable-next-line no-await-in-loop
+            res = await listObjects(usersBucket, { ...listingParams, gt }, log);
+        } catch (error) {
+            log.error('Error during listing', { error });
+            throw error;
+        }
+
+        log.debug('got list of buckets from bucketd', { length: res.length });
+
+        yield res.map(data => {
```

Comment:
We can probably simplify the API and hide the internal paging by changing to:
```suggestion
        yield *res.map(data => {
```

Which would normally yield individual entries instead of the list itself. But since we're processing bucket results in batches afterwards, it may not help as we would have to batch the results anyway.

Reply:
I didn't know about that syntax, thats cool. Originally I had it yielding one bucket at a time but I removed it some I could get the batching downstream.